contractcourt/chain_arbitrator.go

```diff
@@ -1200,5 +1235,98 @@ func (c *ChainArbitrator) SubscribeChannelEvents(
 	return watcher.SubscribeChannelEvents(), nil
 }
 
+// FindOutgoingHTLCDeadline returns the deadline in absolute block height for
+// the specified outgoing HTLC. For an outgoing HTLC, its deadline is defined
+// by the timeout height of its corresponding incoming HTLC - this is the
+// expiry height the that remote peer can spend his/her outgoing HTLC via the
+// timeout path.
+func (c *ChainArbitrator) FindOutgoingHTLCDeadline(scid lnwire.ShortChannelID,
+	outgoingHTLC channeldb.HTLC) fn.Option[int32] {
+
+	// minRefundTimeout tracks the minimal refund timeout found using the
+	// rHash. It's possible that we find multiple HTLCs living in different
+	// channels sharing the same rHash if an MPP is routed by us. In this
+	// case, we'll use the smallest refund timeout as the deadline.
+	//
+	// TODO(yy): can instead query the circuit map to find the exact HTLC.
+	minRefundTimeout := uint32(math.MaxInt32)
+
+	// Find the outgoing HTLC's corresponding incoming HTLC in the circuit
+	// map.
+	rHash := outgoingHTLC.RHash
+	circuit := models.CircuitKey{
+		ChanID: scid,
+		HtlcID: outgoingHTLC.HtlcIndex,
+	}
+	incomingCircuit := c.cfg.QueryIncomingCircuit(circuit)
+
+	// If there's no incoming circuit found, we will use the default
+	// deadline.
+	if incomingCircuit == nil {
+		log.Warnf("ChannelArbitrator(%v): incoming circuit key not "+
+			"found for rHash=%x, using default deadline instead",
+			scid, rHash)
+
+		return fn.None[int32]()
+	}
+
+	// If this is a locally initiated HTLC, it means we are the first hop.
+	// In this case, we can relax the deadline.
+	if incomingCircuit.ChanID.IsDefault() {
+		log.Infof("ChannelArbitrator(%v): using default deadline for "+
+			"locally initiated HTLC for rHash=%x", scid, rHash)
+
+		return fn.None[int32]()
+	}
+
+	log.Debugf("Found incoming circuit %v for rHash=%x using outgoing "+
+		"circuit %v", incomingCircuit, rHash, circuit)
+
+	c.Lock()
+	defer c.Unlock()
+
+	// Iterate over all active channels to find the incoming HTLC specified
+	// by its circuit key.
+	for cp, channelArb := range c.activeChannels {
+		// Skip if the SCID doesn't match.
+		if channelArb.cfg.ShortChanID != incomingCircuit.ChanID {
+			continue
+		}
+
+		// Iterate all the known HTLCs to find the targeted incoming
+		// HTLC.
+		for _, htlcs := range channelArb.activeHTLCs {
+			for _, htlc := range htlcs.incomingHTLCs {
+				// Skip if the index doesn't match.
+				if htlc.HtlcIndex != incomingCircuit.HtlcID {
+					continue
+				}
```

Comment:
IIUC, only one `htlc` should ever match here?  If so, we can short circuit once we find it...

Reply:
oh yeah right, updated.